### PR TITLE
refactor(permissions): introduce Permissions enum

### PR DESCRIPTION
**Pull request by rexledesma**
## Summary
As the title. Raw strings are unwieldy, so let's use enums instead. 

The only gotcha here is to avoid calling `.value` all the time on the enum field, we subclass `str` to duck type the enum members. See https://www.cosmicpython.com/blog/2020-10-27-i-hate-enums.html. However, mypy doesn't recognize this, so we have to cast [1] accordingly.

## Test Plan
pytest

### CODE REVIEW

File: python_modules/dagster/dagster/core/workspace/permissions.py [hunk 1]
```diff
@@ -1,40 +1,63 @@
...
-    "start_sensor": True,
-    "stop_sensor": True,
-    "terminate_pipeline_execution": True,
-    "delete_pipeline_run": True,
-    "reload_repository_location": True,
-    "reload_workspace": True,
-    "wipe_assets": True,
-    "launch_partition_backfill": True,
-    "cancel_partition_backfill": True,
+EDITOR_PERMISSIONS: Dict[str, bool] = {
```
Comment (rexledesma) [inline python_modules/dagster/dagster/core/workspace/permissions.py:46]: [1]

File: python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_permissions.py [hunk 2]
```diff
@@ -40,6 +40,24 @@ def mutate(self, graphene_info, **_kwargs):
...
+    def mutate(self, graphene_info, **_kwargs):
+        pass
+
+
+class FakeMissingEnumPermisisonMutation:
+    @check_permission(Permissions.LAUNCH_PARTITION_BACKFILL)
+    def mutate(self, graphene_info, **_kwargs):
+        pass
+
+
```
Comment (rexledesma) [inline python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_permissions.py:60]: Adding tests as a sanity check on the enum behavior/str subclassing shenanigans

File: python_modules/dagster/dagster/core/workspace/permissions.py (hunk 1 above)
Comment (gibsondan) [reply python_modules/dagster/dagster/core/workspace/permissions.py:46]: probably worth a comment here explaining what you wrote on teh diff

**Review by gibsondan — CHANGES_REQUESTED** [review]
switching the strings to an enum seems good and I support the goal here - but to really get the benefit of an enum don't we want to also change the check_permission and has_permission functions to require that the input be a valid Permisson enum? Otherwise it seems like you could still stick a random string in there and everything would still pass typing-wise.

e.g. that test that allows both strings and Permission enums seems like it should not be possible once we port things from strings to an enum? I may be missing something though

**Review by gibsondan — APPROVED** [review]
eh, this still seems strictly better, we can continue adding more typing and iterating as needed

**Pull request merged**